File: 649.dota-2-参议院.py
#
# @lc app=leetcode.cn id=649 lang=python3
#
# [649] Dota2 参议院
#

# @lc code=start
class Solution:
    def predictPartyVictory(self, senate: str) -> str:
        n = len(senate)
        radiant = collections.deque()   # radiant 队列
        dire = collections.deque()      # dire 队列

        for i, ch in enumerate(senate):  # 记录两方议员的索引
            if ch == "R":   # radiant 入队
                radiant.append(i)
            else:   # dire 入队
                dire.append(i)

        while radiant and dire:
            # 队首比较，如果位次靠前则该议员禁止对方议员参与后续所有投票
            if radiant[0] < dire[0]:    # radiant 禁止 dire
                radiant.append(radiant[0] + n)  # 同时改议员进入队尾等待下次他的投票权的来临
            else:       # dire 禁止 radiant
                dire.append(dire[0] + n)        # 同上
            radiant.popleft()   # 不论如何出队，被禁止则永久出队，未被禁止则去到队尾
            dire.popleft()
        # 最终只会剩下一方议员，判定哪一方不为空即可
        return "Radiant" if radiant else "Dire"

# @lc code=end

File: 649.dota-2-参议院.go
/*
 * @lc app=leetcode.cn id=649 lang=golang
 *
 * [649] Dota2 参议院
 */

// @lc code=start
func predictPartyVictory(senate string) string {
	var radiant, dire []int
	senateLength := len(senate)
	for index, v := range senate {
		if v == 'R' { // radiant 入队
			radiant = append(radiant, index)
		} else { // dire 入队
			dire = append(dire, index)
		}
	}
	// 当双方都仍然有议员在场则就继续新一轮投票
	for len(radiant) > 0 && len(dire) > 0 {
		if radiant[0] < dire[0] { // radiant 议员在前投票，禁止对方队首议员投票权后该议员进入队尾等待下一次投票权
			radiant = append(radiant, radiant[0]+senateLength)
		} else { // dire 议员在前投票，后续同上
			dire = append(dire, dire[0]+senateLength)
		}
		// 双方队首均出列，只不过被禁止的议员永久出队
		radiant = radiant[1:]
		dire = dire[1:]
	}
	// 最后判定双方哪一队不为空
	if len(radiant) > 0 {
		return "Radiant"
	}
	return "Dire"
}

// @lc code=end

